cdc/cdc.cc

```diff
@@ -0,0 +1,221 @@
+// function.
+static future<> populate_desc(db_context ctx, const schema& s) {
+    auto& db = ctx._proxy.get_db().local();
+    auto desc_schema =
+        db.find_schema(s.ks_name(), desc_name(s.cf_name()));
+    auto log_schema =
+        db.find_schema(s.ks_name(), log_name(s.cf_name()));
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    auto ignore_msb_bits =
+        db.get_config().murmur3_partitioner_ignore_msb_bits();
+
+    auto belongs_to = [&](const gms::inet_address& endpoint,
+                          const unsigned int shard_id,
+                          const int shard_count,
+                          const utils::UUID& stream_id) {
+        const auto log_pk = partition_key::from_singular(*log_schema,
+                                                         data_value(stream_id));
+        const auto token = ctx._partitioner.decorate_key(*log_schema, log_pk).token();
+        if (ctx._token_metadata.get_endpoint(ctx._token_metadata.first_token(token)) != endpoint) {
+            return false;
+        }
+        const auto owning_shard_id = dht::murmur3_partitioner::shard_of(
+                token, shard_count, ignore_msb_bits);
+        return owning_shard_id == shard_id;
+    };
+
+    std::vector<mutation> mutations;
+    const auto ts = api::new_timestamp();
+    const auto ck = clustering_key::from_single_value(
+            *desc_schema, timestamp_type->decompose(ts));
+    auto cdef = desc_schema->get_column_definition(to_bytes("stream_id"));
+
+    for_each_endpoint(ctx, [&] (const gms::inet_address& endpoint) {
+        const auto decomposed_ip = inet_addr_type->decompose(endpoint.addr());
+
+        for_each_shard(ctx, endpoint, [&] (unsigned int shard_id, int shard_count) {
+            const auto pk = partition_key::from_exploded(
+                    *desc_schema, { decomposed_ip, int32_type->decompose(static_cast<int>(shard_id)) });
+            mutations.emplace_back(desc_schema, pk);
+
+            auto stream_id = utils::make_random_uuid();
+            while (!belongs_to(endpoint, shard_id, shard_count, stream_id)) {
+                stream_id = utils::make_random_uuid();
+            }
+            auto value = atomic_cell::make_live(*uuid_type,
+                                                ts,
+                                                uuid_type->decompose(stream_id));
+            mutations.back().set_cell(ck, *cdef, std::move(value));
+        });
+    });
+    return ctx._proxy.mutate(std::move(mutations),
+                             db::consistency_level::QUORUM,
+                             db::no_timeout,
+                             nullptr,
+                             empty_service_permit());
+}
+
+future<> setup(db_context ctx, schema_ptr s) {
+    return setup_log(ctx, *s).then([ctx, s = std::move(s)] () mutable {
```

Comment:
For administrative stuff like this it's better to use a seastar::thread. The code is simpler to understand and there's no performance impact.

Reply:
done.